File: src/modules/route.cpp
#include "route.hpp"
#include "../common/common.hpp"
#include <iostream>

int Route::handle_command(const GlobalOptions& options) {
    switch (options.subcommand) {
        case SubCommandType::SHOW:
            return show_routes(options);
        case SubCommandType::ADD:
            return add_route(options);
        case SubCommandType::DELETE:
            return delete_route(options);
        default:
            std::cerr << "Unknown route subcommand" << std::endl;
            return 1;
    }
}

int Route::show_routes(const GlobalOptions& options) {
    Common::log_info("Routing table:");
    std::cout << std::endl;
    
    #ifdef __linux__
    return Common::execute_command("ip", {"route", "show"}, options.dry_run);
    #elif defined(__APPLE__)
    return Common::execute_command("netstat", {"-nr", "-f", "inet"}, options.dry_run);
    #elif defined(_WIN32)
    return Common::execute_command("route", {"print", "-4"}, options.dry_run);
    #endif
    
    return 0;
}

int Route::add_route(const GlobalOptions& options) {
    if (options.args.size() < 1) {
        std::cerr << "Usage: netmgr route add <destination> [--via gateway] [--dev interface]" << std::endl;
        return 1;
    }
    
    std::string destination = options.args[0];
    std::string gateway;
    std::string interface;
    
    // Parse additional arguments
    for (size_t i = 1; i < options.args.size(); i++) {
        if (options.args[i] == "--via" && i + 1 < options.args.size()) {
            gateway = options.args[++i];
        } else if (options.args[i] == "--dev" && i + 1 < options.args.size()) {
            interface = options.args[++i];
        }
    }
    
    Common::log_info("Adding route: " + destination);
    
    #ifdef __linux__
    std::vector<std::string> cmd_args = {"route", "add", destination};
    if (!gateway.empty()) {
        cmd_args.push_back("via");
        cmd_args.push_back(gateway);
    }
    if (!interface.empty()) {
        cmd_args.push_back("dev");
        cmd_args.push_back(interface);
    }
    return Common::execute_command("ip", cmd_args, options.dry_run);
    #elif defined(__APPLE__)
    std::vector<std::string> cmd_args = {"add", "-net", destination};
    if (!gateway.empty()) {
        cmd_args.push_back(gateway);
    }
    return Common::execute_command("route", cmd_args, options.dry_run);
    #elif defined(_WIN32)
    std::vector<std::string> cmd_args = {"add", destination};
    if (!gateway.empty()) {
        cmd_args.push_back("mask");
        cmd_args.push_back("255.255.255.0");
        cmd_args.push_back(gateway);
    }
    return Common::execute_command("route", cmd_args, options.dry_run);
    #endif
    
    return 0;
}

int Route::delete_route(const GlobalOptions& options) {
    if (options.args.empty()) {
        std::cerr << "Usage: netmgr route delete <destination>" << std::endl;
        return 1;
    }
    
    std::string destination = options.args[0];
    Common::log_info("Deleting route: " + destination);
    
    #ifdef __linux__
    return Common::execute_command("ip", {"route", "del", destination}, options.dry_run);
    #elif defined(__APPLE__)
    return Common::execute_command("route", {"delete", destination}, options.dry_run);
    #elif defined(_WIN32)
    return Common::execute_command("route", {"delete", destination}, options.dry_run);
    #endif
    
    return 0;
}

File: src/modules/route.rs
use anyhow::Result;
use crate::cli::RouteCommands;
use crate::common::*;

pub async fn handle_command(command: RouteCommands, options: &GlobalOptions) -> Result<()> {
    match command {
        RouteCommands::Show { route_type } => {
            let rt = route_type.as_deref().unwrap_or("table");
            show_routes(rt, options).await
        }
        RouteCommands::Add { destination, via, dev, table } => {
            add_route(&destination, via.as_deref(), dev.as_deref(), table.as_deref(), options).await
        }
        RouteCommands::Delete { destination, table } => {
            delete_route(&destination, table.as_deref(), options).await
        }
    }
}

async fn show_routes(route_type: &str, options: &GlobalOptions) -> Result<()> {
    match route_type {
        "table" => {
            log_info("Routing table:");
            println!();
            
            #[cfg(target_os = "linux")]
            {
                let output = execute_command("ip", &["route", "show", "table", "main"], options).await?;
                println!("{}", output);
            }
            
            #[cfg(target_os = "windows")]
            {
                let output = execute_command("route", &["print", "-4"], options).await?;
                println!("{}", output);
            }
            
            #[cfg(target_os = "macos")]
            {
                let output = execute_command("netstat", &["-nr", "-f", "inet"], options).await?;
                println!("{}", output);
            }
        }
        "cache" => {
            log_info("Route cache:");
            println!();
            
            #[cfg(target_os = "linux")]
            {
                let output = execute_command("ip", &["route", "show", "cache"], options).await?;
                println!("{}", output);
            }
            
            #[cfg(not(target_os = "linux"))]
            {
                log_info("Route cache display not supported on this platform");
            }
        }
        "all" => {
            log_info("All routing tables:");
            println!();
            
            #[cfg(target_os = "linux")]
            {
                let output = execute_command("ip", &["route", "show", "table", "all"], options).await?;
                println!("{}", output);
            }
            
            #[cfg(target_os = "windows")]
            {
                let output = execute_command("route", &["print"], options).await?;
                println!("{}", output);
            }
            
            #[cfg(target_os = "macos")]
            {
                let output = execute_command("netstat", &["-nr"], options).await?;
                println!("{}", output);
            }
        }
        _ => {
            log_info(&format!("Routes for {}:", route_type));
            println!();
            
            #[cfg(target_os = "linux")]
            {
                let output = execute_command("ip", &["route", "show", route_type], options).await?;
                println!("{}", output);
            }
            
            #[cfg(not(target_os = "linux"))]
            {
                return Err(NetMgrError::UnsupportedPlatform("Showing specific routes not implemented for this platform".to_string()).into());
            }
        }
    }
    
    Ok(())
}

async fn add_route(destination: &str, via: Option<&str>, dev: Option<&str>, table: Option<&str>, options: &GlobalOptions) -> Result<()> {
    log_info(&format!("Adding route: {}", destination));
    
    #[cfg(target_os = "linux")]
    {
        let mut args = vec!["route", "add", destination];
        if let Some(gateway) = via {
            args.extend(&["via", gateway]);
        }
        if let Some(device) = dev {
            args.extend(&["dev", device]);
        }
        if let Some(tbl) = table {
            if tbl != "main" {
                args.extend(&["table", tbl]);
            }
        }
        
        execute_command("ip", &args, options).await?;
    }
    
    #[cfg(target_os = "windows")]
    {
        let mut args = vec!["add", destination];
        if let Some(gateway) = via {
            args.extend(&["mask", "255.255.255.0", gateway]);
        }
        if let Some(device) = dev {
            args.extend(&["if", device]);
        }
        
        execute_command("route", &args, options).await?;
    }
    
    #[cfg(target_os = "macos")]
    {
        let mut args = vec!["add", "-net", destination];
        if let Some(gateway) = via {
            args.push(gateway);
        }
        
        execute_command("route", &args, options).await?;
    }
    
    Ok(())
}

async fn delete_route(destination: &str, table: Option<&str>, options: &GlobalOptions) -> Result<()> {
    log_info(&format!("Deleting route: {}", destination));
    
    #[cfg(target_os = "linux")]
    {
        let mut args = vec!["route", "del", destination];
        if let Some(tbl) = table {
            if tbl != "main" {
                args.extend(&["table", tbl]);
            }
        }
        
        execute_command("ip", &args, options).await?;
    }
    
    #[cfg(target_os = "windows")]
    {
        execute_command("route", &["delete", destination], options).await?;
    }
    
    #[cfg(target_os = "macos")]
    {
        execute_command("route", &["delete", destination], options).await?;
    }
    
    Ok(())
}
